### PR TITLE
NIR: asm statement support, Wip

### PR DESCRIPTION
**Pull request by ASVIEST**
`GlobalAsm` instr is basic asm, which is just str literals or some instrs

`Asm` instr is GCC extended asm
https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
```
Asm {
  AsmTemplate {
     Some asm code
     SymUse nimInlineVar # `a`
     Some asm code
   }
  AsmOutputOperand {
     # [asmSymbolicName] constraint (nimVariableName)
     AsmInjectExpr {symUse nimVariableName} # for output it have only one sym (lvalue)
     asmSymbolicName # default: ""
     constraint
  }
  AsmInputOperand {
     # [asmSymbolicName] constraint (nimExpr)
     AsmInjectExpr {symUse nimVariableName} # (rvalue)
     asmSymbolicName # default: ""
     constraint
  }
  AsmClobber {
    "clobber"
  }
}
```
This structure is simple for codegens (including those what may be in the future like libgccjit or LLVM).
Also it can produce some nim errors without dependencies of backends.

### CODE REVIEW

File: compiler/nir/nirinsts.nim [hunk 1]
```diff
@@ -108,6 +108,15 @@ type
...
     Emit,
+
+    AsmGlobal # top-level basic asm. Better to use extended asm if it possible
+    Asm, # extended asm
+    AsmTemplate,
+    AsmInjectExpr, # like cexpr in gcc
+    AsmOutputOperand,
+    AsmInputOperand,
+    AsmClobber,
+
```
Comment (Araq) [inline compiler/nir/nirinsts.nim:116]: Too GCC specific, should use `Emit` as the umbrella opcode. `Emit` having a list of `Verbatim` and `SymUse` etc opcodes. But it needs a previously missing `Verbatim` opcode that takes a string literal like `StrVal`.

Comment (Araq) [reply compiler/nir/nirinsts.nim:116]: Probably also needs new pragmas that specify which section the Emit refers to.

Comment (ASVIEST) [reply compiler/nir/nirinsts.nim:116]: > Too GCC specific, should use `Emit` as the umbrella opcode. `Emit` having a list of `Verbatim` and `SymUse` etc opcodes. But it needs a previously missing `Verbatim` opcode that takes a string literal like `StrVal`.


1. I understand correctly that asm stategments should be in Emit and it is universal for asembler and emitting code ? Then probably we should enter an enum with all uses and get it from Emit. For example:
```
Emit {
  EmitTarget "Asm" # or EmitTarget "EmitCode"
}.
```
If so, that's really cool.
2. What is Verbatim doing ? I hope not for this kind of code:
```
Emit {
  Verbatim "AsmTemplate"
  StrVal "
   mov %1, %0\n
   add $1, %0\n
  "
  Verbatim "OutputOperand"
  SymUse 3 #result var
  StrVal "" #no asm symbolic name
  StrVal "=r"
  ...
}.
```
This is tricky for code generation, perhaps this is better:
```
Emit {
  Verbatim "AsmTemplate" {
    StrVal "
      mov %1, %0
      add $1, %0\n
    "
  }
  Verbatim "OutputOperand" {
    SymUse 3 #result var
    StrVal "" #no asm symbolic name
    StrVal "=r"
  }
  ...
}
```
And maybe allowed verbatim values must be enum.

Comment (Araq) [reply compiler/nir/nirinsts.nim:116]: Verbatim is a raw string literal, that means it's copied into the C code output buffer without any quoting. It is not an obfuscated enum. It's useful as Visual C++'s `asm` statement doesn't use quotes, for example.

**Comment by Araq** [comment]
Btw here is my NIR todo:

# NIR TODO:
- Implement closure calls.
- Port block exiting logic from the C codegen.
- Port parameter passing and NRVO.
- Port 3 opcodes to enable closure iterators.
- Port Runtime type information from the C codegen.
- Port the few missing magics from the C codegen.

**Comment by ASVIEST** [comment]
```nim
# asm (for gcc extended asm it includes also AsmTemplate, AsmOutOperand*, AsmInOperand*)
Emit {
  EmitTarget "Asm"
  Info {
    InfoKind "IsGlobal"
    false
  }

  # only GCC extended asm info.
  # It can be used by backend or not used.
  Info {
    InfoKind "AsmTemplate"
    Verbatim """
      mov %1, %0
      add $1, %0
    """
  }
  # there can be any number of operands
  Info {
   InfoKind "AsmOutOperand"
   SymUse r
   Verbatim ""
   Verbatim "=r"
  }
  Info {
   InfoKind "AsmInOperand"
   SymUse r
   Verbatim ""
   Verbatim "r"
  }

  EmitCode {
    Verbatim """
      mov %1, %0
      add $1, %0 
      :"=r"(
    """
    SymUse result
    Verbatim """
      )
      :"r"(
    """
    SymUse dst
    Verbatim ")"
   """
  }
```
When using gcc extended asm, it contains useful information for backends.
The EmitCode instruction directly emits code that is a list of Verbatim, SymUse, etc. into c code (as an example).

What do you think about this design ?

**Comment by Araq** [comment]
> What do you think about this design ?

It's wrong. You effectively do the GCC specific asm parsing in the Nim frontend, but it belongs into the "NIR to C" backend part. Only the Nim specific parsing should be done in the frontend.

File: compiler/nir/nirinsts.nim [hunk 2]
```diff
@@ -245,6 +252,11 @@ proc nextChild(tree: Tree; pos: var int) {.inline.} =
 proc next*(tree: Tree; pos: var NodePos) {.inline.} = nextChild tree, int(pos)
 
 template firstSon*(n: NodePos): NodePos = NodePos(n.int+1)
+proc lastSon*(tree: Tree; n: NodePos): NodePos =
```
Comment (Araq) [inline compiler/nir/nirinsts.nim]: That's not how to compute the last son.

**Comment by Araq** [comment]
Already 600 lines that accomplishes what exactly? Parsing the clobber lists of GCC's unreadable inline assembler in order to pass it back to GCC to parse it for itself. Sorry, but this is terrible.

**Comment by ASVIEST** [comment]
The initial goal was to help future backends like LLVM or libgccjit in parsing the extended assembler (as auxiliary information) when it is available, since there is no parsing of the assembler itself. It is more logical to do it not on the frontend, but perhaps on an intermediate stage before the backend (Nir -> Nir) that has information about basic things, such as for example assembler syntax (gcc or Visual C++), or on the backend, which is worse. Asm parsing is mainly aimed at this future backends(LLVM, libgccjit), but it have some pluses on other targets (C, C++).

The main plus for C(++) purposes is that the parsing is done on the Nim side, not the compiler, so it can then be parsed, e.g. not allowing such code
```nim
proc test(a: int64) =
  asm """
    add %[val], %[a]
    :[a]"=r"(`a`)
    :[val]"r"((long long)(4/2))
  """

var a: int64 = 35
test(a)
echo a
```
This code compiles (although it shouldn't because it tries to change a parameter), but it can't change the variable a (it stays with the value 35), which is good, but the fact that nim's code is compiled (with invalid behavour) is not pleasant.
```nim
proc test(a: var int64) =
  asm """
    add %[val], %[a]
    :[a]"=r"(`a`)
    :[val]"r"((long long)(4/2))
  """

var a: int64 = 35
test(a)
echo a
```
Here the code compiles again, but again it doesn't change a (for C through GCC), or when compiling in C++ through GCC, a becomes random number.
Because of this, you should write this code instead of the previous one:
```nim
proc test(a: var int64) =
  var b = a
  asm """
    add %[val], %[a]
    :[a]"=r"(`b`)
    :[val]"r"((long long)(4/2))
  """
  a = b

var a: int64 = 35
test(a)
echo a 
```
Now it works as it should.

Analysis of gcc extended asm also move some C, C++ compiler exceptions to Nir, which means they can be get in nimsuggest, etc.

Without such analysis, adding support for such cases is simply *not possible* because the Nim compiler will think of assembly language as just a black box, not knowing what is being changed, etc. 
About repetitive parsing of assembler in GCC(or clang) and in Nim is completely true, but it is the simplest thing in all inline asm, the main complexity of it in the interaction of the built-in assembler with the compiler-generated assembler and code, for example, a jump in the assembler should not violate the space of compiler-generated labels, etc. 
See https://github.com/bytecodealliance/wasmtime/issues/1041#issuecomment-531574394.
It's really hard work, but not parsing.

**Comment by Araq** [comment]
Ok, well, currently not even "hello world" works with NIR, so can we focus on the essentials please before adding superior logic for inline asm...

**Comment by ASVIEST** [comment]
Ok, then I'll do it without parsing and analyzing the inline asm, for now, and keep the parsing part in a separate file for when NIR is more complete and there's time to add support for assembler analyzing.

**Comment by arnetheduck** [comment]
fwiw, I wanted to do this for [nlvm](https://github.com/arnetheduck/nlvm) which would benefit from proper asm parsing (so it can set up the correct constraints like gcc does from the nim-inlined asm code) - if the asm parser can be separate, that would certainly be helpful

**Comment by ASVIEST** [comment]
You can't just import. Keep in mind that this only works with NIR instructions (previously with ast, but now it is not supported), and not with strings, since the asm stategment may contain nim symbols and expressions. Therefore, it will be more correct to use this when nlvm generates llvm IR from NIR, and not from ast. For now you can take gcc_extended_asm.nim and replace it's tokenizer by genasm.nim tokenizer (parseAsm iterator). Note: it don't parses labels.

File: compiler/nir/nirinsts.nim [hunk 3]
```diff
@@ -254,10 +254,17 @@ proc next*(tree: Tree; pos: var NodePos) {.inline.} = nextChild tree, int(pos)
...
+    oldPos = pos
+
   assert tree.nodes[pos].kind > LastAtomicValue
   let last = pos + tree.nodes[pos].rawSpan
-  NodePos last
+  inc pos
+  while pos < last:
+    oldPos = pos
+    nextChild tree, pos
+  NodePos oldPos
```
Comment (Araq) [inline compiler/nir/nirinsts.nim]: The algorithm for last son is like this:

```nim

proc parentImpl(tree: Tree; n: NodePos): NodePos =
  # finding the parent of a node is rather easy:
  var pos = n.int - 1
  while pos >= 0 and isAtom(tree, pos) or (pos + tree.nodes[pos].rawSpan - 1 < n.int):
    dec pos
  assert pos >= 0, "node has no parent"
  result = NodePos(pos)

proc isLastSon*(tree: Tree; parent, n: NodePos): bool {.inline.} =
  # A node is a the last son of a parent node if its span
  # falls onto the end of the parent's span:
  let last = n.int + span(tree, n.int)
  result = last == parent.int + span(tree, parent.int)

proc lastSon*(tree: Tree; n: NodePos): NodePos =
  assert(not isAtom(tree, n.int))
  result = NodePos(n.int + span(tree, n.int) - 1)
  while true:
    let p = parentImpl(tree, result)
    if p == n: break
    result = p
```

File: compiler/nir/nirtypes.nim [hunk 4]
```diff
@@ -62,6 +62,7 @@ proc hash*(a: TypeId): Hash {.borrow.}
 type
   Literals* = ref object
     strings*: BiTable[string]
+    verbatims*: BiTable[string]
```
Comment (Araq) [inline compiler/nir/nirtypes.nim]: Completely unnecessary, just use the existing `strings`.

**Comment by ASVIEST** [comment]
Currently this can generate code for GCC and VCC inline asm, this can be selected via the `-a gcc-like`(or `msvc-like`) flag on nirc. The basic assembler also works. noAsmStackFrame also works, instead of generate __declspec or attribute naked it generates NIM_NAKED C macro. Assembler parsing is separated into a separated file, which is not currently used. Assembly analysis will be offered as rfc.

RFC: https://github.com/nim-lang/RFCs/issues/542

**Comment by Araq** [comment]
> Currently this can generate code for GCC and VCC inline asm, this can be selected via the -a gcc-like(or msvc-like) flag on nirc.

The used inline assembler format should be a pragma though, not a command line option and `nimrc` should have a `dialect` switch so that we can say "produce C code compatible with VCC etc".

**Comment by ASVIEST** [comment]
> > Currently this can generate code for GCC and VCC inline asm, this can be selected via the -a gcc-like(or msvc-like) flag on nirc.
> 
> The used inline assembler format should be a pragma though, not a command line option and `nimrc` should have a `dialect` switch so that we can say "produce C code compatible with VCC etc".


-a flag say "can produce C code from inline asm statements  compatible with VCC etc".
If there will be only dialect switch (without -a flag), what dialect will be for example ICC (intel C compiler) ?  How dialect can show that C compiler don't support inline asm (like some tcc builds) ? If for ICC etc. it is a different dialect, then the codegen will have to check all dialects, etc. I think it is better to leave the -a flag for customization and add the targetCC (or dialect) flag, which sets the default settings for the selected compiler.

The pragma idea is good for deciding what syntax to generate assembler code with, for example for ICC, the assembler can be either gcc style or vcc style. If code uses an procedure with assembler from an external library, the assembler stmt must have a pragma that explains it gcc(AT&T) or vcc(intel), then it can use libraries that that use gcc or vcc syntax.

**Comment by ASVIEST** [comment]
Now it can resolve asm syntax kind. For example backend Inline asm syntax is {"gcc", "vcc"} and via `{.inlineAsmSyntax: "gcc".}` can specify that asm stmt with gcc syntax. 
NOTE: this pragma not required when CC support only one syntax: gcc/vcc, but recommend to make code more universal.

File: compiler/nir/gcc_extended_asm.nim [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    AsmTemplate
+    AsmOutputOperand
+    AsmInputOperand
+    AsmClobber
+    AsmGotoLabel
+
+    AsmInjectExpr
+    AsmStrVal
+
+  GccAsmNode* = ref object
```
Comment (Araq) [inline compiler/nir/gcc_extended_asm.nim:78]: That's not how we write ASTs in Nim 2023. We now use "packed trees". The files you have touched can guide you.

File: compiler/nir/ast2ir.nim [hunk 6]
```diff
@@ -2517,6 +2518,53 @@ proc genComplexCall(c: var ProcCon; n: PNode; d: var Value) =
...
 
+template genEmitCode(c: var ProcCon; n: PNode; isAsmStmt: bool) =
+  let offset =
+    if isAsmStmt: 1 # first son is pragmas
+    else: 0
+  
+  build c.code, info, EmitCode:
+    for i in offset..<n.len:
+      let it = n[i]
+      case it.kind:
```
Comment (Araq) [inline compiler/nir/ast2ir.nim:2529]: That's not how the rest of the codebase indents the `case` statement.

File: compiler/nir/cir.nim [hunk 7]
```diff
@@ -31,6 +32,7 @@ type
     Semicolon = ";"
     Comma = ", "
     Space = " "
+    Quote = $'"'
```
Comment (Araq) [inline compiler/nir/cir.nim:35]:
```suggestion
    Quote = "\""
```

File: compiler/nir/cir.nim [hunk 8]
```diff
@@ -554,6 +562,54 @@ template moveToDataSection(body: untyped) =
...
+    result
+
+  maybeAddQuote
+  for ch in sons(t, n):
+    if t[ch].kind == Verbatim:
+      s = c.m.lit.strings[t[ch].litId]
+      left = 0
+      for i in 0..s.high:
+        if s[i] == '\n':
+          c.add s[left..i - 1]
```
Comment (Araq) [inline compiler/nir/cir.nim:591]: What? Why? Didn't you write a parser for the asm? Shouldn't it split it at newlines if they are important?

File: compiler/nir/cir.nim [hunk 9]
```diff
@@ -554,6 +562,54 @@ template moveToDataSection(body: untyped) =
...
   setLen c.code, oldLen
 
+proc genGccAsm(c: var GeneratedCode; t: Tree; n: NodePos) =
+  c.add "__asm__ "
+  c.add ParLe
+  c.add NewLine
+
+  var asmTemplate = true
+  var left = 0
+  var s = ""
```
Comment (Araq) [inline compiler/nir/cir.nim:572]: Why is `s` mutable?

File: compiler/nir/cir.nim [hunk 10]
```diff
@@ -807,11 +863,65 @@ proc gen(c: var GeneratedCode; t: Tree; n: NodePos) =
...
+
+        if not isBasicAsm:
+          if inlineAsmSyntax == {GCCExtendedAsm}: genGccAsm(c, t, code)
+          elif inlineAsmSyntax == {VisualCPP}: genVisualCPPAsm(c, t, code)
+          else: raiseAssert "Not implemented inline asm syntax"
+        else:
+          assert (
+            isLastSon(t, code, code.firstSon) and
+            t[code.firstSon].kind == Verbatim
+          ), "Invalid basic asm. Basic asm should be only one verbatim"
```
Comment (Araq) [inline compiler/nir/cir.nim:910]: Shouldn't the frontend do this?

File: compiler/nir/cir.nim [hunk 11]
```diff
@@ -807,11 +863,65 @@ proc gen(c: var GeneratedCode; t: Tree; n: NodePos) =
...
+          genGccAsm(c, t, code)
+
+      of Code:
+        raiseAssert"not supported"
+
+  of EmitTarget:
```
Comment (Araq) [inline compiler/nir/cir.nim:916]: Why is a EmitTarget an opcode of its own? We have a pragma annotation system in place for this...

**Comment by Araq** [comment]
700 lines we have to maintain for good all the while "NIR" cannot even do "hello world". Enough of this.

**Comment by Araq** [comment]
Reopen it once NIR can do something useful at all.

**Pull request closed without merging**